### PR TITLE
windows: Fix remaining compiler warnings (MSVC)

### PR DESCRIPTION
**Pull request by practicalswift**
Fix remaining compiler warnings (MSVC).

Before:

```
$ msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nowarn:C4244;C4267;C4715 /nologo
…\script\script.cpp(272): warning C4018: '>': signed/unsigned mismatch
…\test\allocator_tests.cpp(147): warning C4312: 'reinterpret_cast': conversion from 'int' to 'void *' of greater size
…\boost\test\tools\old\impl.hpp(107): warning C4805: '==': unsafe mix of type 'const Left' and type 'const Right' in operation
…\test\crypto_tests.cpp(535): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
…\test\script_tests.cpp(188): warning C4805: '==': unsafe mix of type 'int' and type 'bool' in operation
…\test\script_tests.cpp(190): warning C4805: '==': unsafe mix of type 'int' and type 'bool' in operation
…\test\script_tests.cpp(191): warning C4805: '==': unsafe mix of type 'int' and type 'bool' in operation
$
```

After:

```
$ msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nowarn:C4244;C4267;C4715;C4805 /nologo
$
```

### CODE REVIEW

File: src/script/script.cpp [hunk 1]
```diff
@@ -269,7 +269,7 @@ bool CScript::HasValidOps() const
     while (it < end()) {
         opcodetype opcode;
         std::vector<unsigned char> item;
-        if (!GetOp(it, opcode, item) || opcode > MAX_OPCODE || item.size() > MAX_SCRIPT_ELEMENT_SIZE) {
+        if (!GetOp(it, opcode, item) || opcode > MAX_OPCODE || item.size() > (size_t)MAX_SCRIPT_ELEMENT_SIZE) {
```
Comment (MarcoFalke) [inline src/script/script.cpp]: `MAX_SCRIPT_ELEMENT_SIZE` is  of type `unsigned`, so I don't see why this change is here.

Comment (practicalswift) [reply src/script/script.cpp]: Sorry, should have been `(unsigned int)opcode`. Now updated. Please re-review :-)

**Comment by Empact** [comment]
How about resolving opcodetype discord like so:
https://github.com/Empact/bitcoin/commit/76561599271bda0cfbb7a9a43bb69d01509dd747.patch

File: src/script/script.cpp [hunk 2]
```diff
@@ -269,7 +269,7 @@ bool CScript::HasValidOps() const
     while (it < end()) {
         opcodetype opcode;
         std::vector<unsigned char> item;
-        if (!GetOp(it, opcode, item) || opcode > MAX_OPCODE || item.size() > MAX_SCRIPT_ELEMENT_SIZE) {
+        if (!GetOp(it, opcode, item) || (unsigned int)opcode > MAX_OPCODE || item.size() > MAX_SCRIPT_ELEMENT_SIZE) {
```
Comment (scravy) [inline src/script/script.cpp]: Nit: could be a `static_cast`

File: src/test/crypto_tests.cpp [hunk 3]
```diff
@@ -532,7 +532,7 @@ BOOST_AUTO_TEST_CASE(countbits_tests)
             // Check handling of zero.
             BOOST_CHECK_EQUAL(CountBits(0), 0U);
         } else if (i < 10) {
-            for (uint64_t j = 1 << (i - 1); (j >> i) == 0; ++j) {
+            for (uint64_t j = ((uint64_t) 1) << (i - 1); (j >> i) == 0; ++j) {
```
Comment (scravy) [inline src/test/crypto_tests.cpp]: Nit: could be a `static_cast`

**Comment by sipsorcery** [comment]
tACK https://github.com/bitcoin/bitcoin/pull/14151/commits/e99684c84094d5f04382d91eed00ec3ef911fd03

Prior to this PR:

```
msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nowarn:C4244;C4267;C4715 /nologo
c:\projects\bitcoin-72c17\src\script\script.cpp(272): warning C4018: '>': signed/unsigned mismatch [C:\projects\bitcoin-72c17\build_msvc\libbitcoinconsensus\libbitcoinconsensus.vcxproj]
c:\projects\bitcoin-72c17\src\test\allocator_tests.cpp(147): warning C4312: 'reinterpret_cast': conversion from 'int' to 'void *' of greater size [C:\projects\bitcoin-72c17\build_msvc\test_bitcoin\test_bitcoin.vcxproj]
c:\tools\vcpkg\installed\x64-windows-static\include\boost\test\tools\old\impl.hpp(107): warning C4805: '==': unsafe mix of type 'const Left' and type 'const Right' in operation [C:\projects\bitcoin-72c17\build_msvc\test_bitcoin\test_bitcoin.vcxproj]
c:\projects\bitcoin-72c17\src\test\crypto_tests.cpp(535): warning C4334: '<<': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?) [C:\projects\bitcoin-72c17\build_msvc\test_bitcoin\test_bitcoin.vcxproj]
c:\projects\bitcoin-72c17\src\test\script_tests.cpp(188): warning C4805: '==': unsafe mix of type 'int' and type 'bool' in operation [C:\projects\bitcoin-72c17\build_msvc\test_bitcoin\test_bitcoin.vcxproj]
c:\projects\bitcoin-72c17\src\test\script_tests.cpp(190): warning C4805: '==': unsafe mix of type 'int' and type 'bool' in operation [C:\projects\bitcoin-72c17\build_msvc\test_bitcoin\test_bitcoin.vcxproj]
c:\projects\bitcoin-72c17\src\test\script_tests.cpp(191): warning C4805: '==': unsafe mix of type 'int' and type 'bool' in operation [C:\projects\bitcoin-72c17\build_msvc\test_bitcoin\test_bitcoin.vcxproj]
move build_msvc\%PLATFORM%\%CONFIGURATION%\*.iobj build_msvc\cache\
```

Subsequently:

```
msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nowarn:C4244;C4267;C4715;C4805 /nologo
move build_msvc\%PLATFORM%\%CONFIGURATION%\*.iobj build_msvc\cache\
```

File: src/test/allocator_tests.cpp [hunk 4]
```diff
@@ -144,6 +144,10 @@ class TestLockedPageAllocator: public LockedPageAllocator
                 *lockingSuccess = true;
             }
 
+#ifdef _MSC_VER
+// Disable MSVC warning: C4312: 'reinterpret_cast': conversion from 'int' to 'void *' of greater size
+#pragma warning(disable: 4312)
+#endif
```
Comment (MarcoFalke) [inline src/test/allocator_tests.cpp]: Can't this be solved by `return reinterpret_cast<void*>(uint64_t{0x08000000 + (count<<24)});`?

**Comment by practicalswift** [comment]
@MarcoFalke Please re-review :-)

**Comment by ken2812221** [comment]
utACK 94ce7c9a351127c5972e4192d9934c10d2065c4b

**Comment by practicalswift** [comment]
@ken2812221 Squashed. Please re-review :-)

**Comment by ken2812221** [comment]
utACK 224cc41

File: appveyor.yml [hunk 5]
```diff
@@ -35,7 +35,7 @@ before_build:
        }
 - ps:  Start-Process clcache-server
 build_script:
-- cmd: msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nowarn:C4244;C4267;C4715 /nologo
+- cmd: msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nowarn:C4244;C4267;C4715;C4805 /nologo
```
Comment (NicolasDorier) [inline appveyor.yml]: I think instead of bundling that in /nowarn we should include the nowarn inside the generated project file.

Comment (NicolasDorier) [reply appveyor.yml]: `<PropertyGroup><NoWarn>C4244;C4267;C4715;C4805</NoWarn></PropertyGroup>`

Comment (practicalswift) [reply appveyor.yml]: @NicolasDorier I'll fix that! I'm not really familiar with the MSVC project files - in which of them should that be included?

These are the files with `PropertyGroup`:

```
$ git grep -l '<PropertyGroup'
build_msvc/bench_bitcoin/bench_bitcoin.vcxproj
build_msvc/bitcoin-cli/bitcoin-cli.vcxproj
build_msvc/bitcoin-tx/bitcoin-tx.vcxproj
build_msvc/bitcoind/bitcoind.vcxproj
build_msvc/common.init.vcxproj
build_msvc/common.vcxproj
build_msvc/libbitcoin_cli/libbitcoin_cli.vcxproj.in
build_msvc/libbitcoin_common/libbitcoin_common.vcxproj.in
build_msvc/libbitcoin_crypto/libbitcoin_crypto.vcxproj.in
build_msvc/libbitcoin_qt/libbitcoin_qt.vcxproj
build_msvc/libbitcoin_server/libbitcoin_server.vcxproj.in
build_msvc/libbitcoin_util/libbitcoin_util.vcxproj.in
build_msvc/libbitcoin_wallet/libbitcoin_wallet.vcxproj.in
build_msvc/libbitcoin_zmq/libbitcoin_zmq.vcxproj.in
build_msvc/libbitcoinconsensus/libbitcoinconsensus.vcxproj
build_msvc/libunivalue/libunivalue.vcxproj
build_msvc/test_bitcoin-qt/test_bitcoin-qt.vcxproj
build_msvc/test_bitcoin/test_bitcoin.vcxproj
build_msvc/testconsensus/testconsensus.vcxproj
```

Comment (practicalswift) [reply appveyor.yml]: Tried this but it didn't suppress the warnings:

```diff
diff --git a/appveyor.yml b/appveyor.yml
index 147f458d9..ae6bd2f89 100644
--- a/appveyor.yml
+++ b/appveyor.yml
@@ -35,7 +35,7 @@ before_build:
        }
 - ps:  Start-Process clcache-server
 build_script:
-- cmd: msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nowarn:C4244;C4267;C4715;C4805 /nologo
+- cmd: msbuild /p:TrackFileAccess=false /p:CLToolExe=clcache.exe build_msvc\bitcoin.sln /m /v:q /nologo
 after_build:
 - cmd: move build_msvc\%PLATFORM%\%CONFIGURATION%\*.iobj build_msvc\cache\
 - cmd: move build_msvc\%PLATFORM%\%CONFIGURATION%\*.ipdb build_msvc\cache\
diff --git a/build_msvc/common.init.vcxproj b/build_msvc/common.init.vcxproj
index c3c0f665c..4ba2aca6c 100644
--- a/build_msvc/common.vcxproj
+++ b/build_msvc/common.vcxproj
@@ -7,9 +7,12 @@
           $(BuildDependsOn);
       </BuildDependsOn>
   </PropertyGroup>
+  <PropertyGroup>
+      <NoWarn>C4244;C4267;C4715;C4805</NoWarn>
+  </PropertyGroup>
   <Target Name="CopyConfig"
           Inputs="$(MSBuildThisFileDirectory)bitcoin_config.h"
           Outputs="$(MSBuildThisFileDirectory)..\src\config\bitcoin-config.h">
       <Copy SourceFiles="$(MSBuildThisFileDirectory)bitcoin_config.h" DestinationFiles="$(MSBuildThisFileDirectory)..\src\config\bitcoin-config.h" />
   </Target>
```

Comment (NicolasDorier) [reply appveyor.yml]: You normally did right. I take a look.

**Comment by NicolasDorier** [comment]
Adding `/warnaserror` is a bit too much?

**Comment by NicolasDorier** [comment]
For test, I tried running a dry compile with `/warnaserror`, it worked fine but two warning show up. (If I recompile a second time, without cleaning, it does not show)

```powershell
"C:\Sources\bitcoin\build_msvc\bitcoin.sln" (default target) (1) ->
"C:\Sources\bitcoin\build_msvc\libbitcoinconsensus\libbitcoinconsensus.vcxproj" (default target) (2) ->
(ClCompile target) ->
  c:\program files (x86)\microsoft visual studio\2017\buildtools\vc\tools\msvc\14.14.26428\include\xutility(3182): error C4996: 'std::equal::_Unchecked_iterators::_Deprecate': Call to 'st
d::equal' with parameters that may be unsafe - this call relies on the caller to check that the passed values are correct. To disable this warning, use -D_SCL_SECURE_NO_WARNINGS. See docu
mentation on how to use Visual C++ 'Checked Iterators' [C:\Sources\bitcoin\build_msvc\libbitcoinconsensus\libbitcoinconsensus.vcxproj]


"C:\Sources\bitcoin\build_msvc\bitcoin.sln" (default target) (1) ->
"C:\Sources\bitcoin\build_msvc\testconsensus\testconsensus.vcxproj" (default target) (3) ->
"C:\Sources\bitcoin\build_msvc\libbitcoin_util\libbitcoin_util.vcxproj" (default target) (4) ->
(Lib target) ->
  sync.obj : error LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library [C:\Sources\b
itcoin\build_msvc\libbitcoin_util\libbitcoin_util.vcxproj]
```
That said, I don't think this is very important,

tACK 224cc41bb9fd2906d9a61abd0cad9bd77365569b

**Comment by practicalswift** [comment]
@NicolasDorier Personally I'm all for `/warnaserror` assuming AppVeyor build failures don't block merge (and thus get in the way of the maintainers job). They don't right?

**Comment by NicolasDorier** [comment]
The whole purpose of `/warnaserror` is to break the build. I tried `/warnaserror`, and there is the above error for the first build. I don't think it is worth fixing/fixable.

**Comment by practicalswift** [comment]
@NicolasDorier Yes, I know but I thought the project was configured to block merge in case of Travis build failure but not in case of AppVeyor build failure. But I'm probably mistaken :-)

**Comment by NicolasDorier** [comment]
ah yes I think you are right. Well, I think we can ignore `/warnaserror` for now, except if you find an obvious fix.

**Comment by sipsorcery** [comment]
For a C++ build to exclude specific warnings and set the "treat warnings as error" flag the following node can be added to the common.vcxproj file:

```
<Project>
  <ItemDefinitionGroup>
	<ClCompile>
	  <DisableSpecificWarnings>4244;4267;4715;4805;</DisableSpecificWarnings>
	  <TreatWarningAsError>true</TreatWarningAsError>
	</ClCompile>
  </ItemDefinitionGroup>
</Project>
```

When warnings as error is set the script\interpreter.cpp code fails to compile on line 264 and specifically this check:

```std::equal(b.begin(), b.end(), pc)```

A quick Google found a blog post about Visual C++ [range checking safety](https://blogs.msdn.microsoft.com/vcblog/2017/05/05/why-am-i-getting-these-_scl_secure_no_warnings-messages/). To ignore this warning add `4996` to the list of specific warnings and the solution will then compile.

**Comment by sipsorcery** [comment]
tACK https://github.com/bitcoin/bitcoin/pull/14151/commits/f14f4a68850947f04ab6f94ec6e52a07fb1cd61f.

**Comment by ken2812221** [comment]
re-utACK f14f4a68

**Comment by practicalswift** [comment]
Rebased!

**Comment by practicalswift** [comment]
@sipsorcery @ken2812221 @MarcoFalke Thanks for reviewing. Would you mind re-reviewing? :-)

**Comment by ken2812221** [comment]
utACK 26a9c9e4ab225cdb276ff8bfde4b09fd939a3bae

File: src/test/script_tests.cpp [hunk 6]
```diff
@@ -184,11 +184,12 @@ void DoTest(const CScript& scriptPubKey, const CScript& scriptSig, const CScript
     stream << tx2;
     int libconsensus_flags = flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_ALL;
     if (libconsensus_flags == flags) {
+        int expectedSuccessCode = expect ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/script_tests.cpp]: No, we want `expectedSuccessCode` to be `1` (if `expect`) or `0` (if `!expect`) :-)

[the remark this answers is not in the document]
Comment (ken2812221) [reply src/test/script_tests.cpp]: This is correct, the integer value is assigned to expectedSuccessCode.

**Comment by NicolasDorier** [comment]
ping @MarcoFalke, anything missing for merge?

**Comment by MarcoFalke** [comment]
Generally we don't touch consensus code to silence compiler warnings of some specific compiler version, since there is (to the best of my knowledge) no method yet to review the changes for correctness. Reviewers well-versed with compiler internals and language specification could look for issues, but a formal method to prove correctness is currently not existing.

**Comment by practicalswift** [comment]
@MarcoFalke Now suppressing the warning in `src/script/script.cpp` instead. Would you mind re-reviewing? :-)

**Comment by ken2812221** [comment]
re-utACK fc2e8c0

**Comment by practicalswift** [comment]
Rebased! :-)

**Comment by NicolasDorier** [comment]
reACK, can it be merged? This has been ACK several time already.

**Comment by practicalswift** [comment]
Can I do anything to increase the likelihood of getting this PR merge? :-)

**Comment by NicolasDorier** [comment]
Dance on one foot clapping your hands singing praise to the genesis block.

**Comment by ken2812221** [comment]
ACK 896eddb28aae304d8f380daba5abb38fe134e49e

**Comment by Empact** [comment]
nit: somewhat prefer `ull` postfix over `static_cast<uint64_t>(literal)`

utACK 896eddb

**Comment by practicalswift** [comment]
@MarcoFalke Would you mind re-reviewing? Let me know if I can do anything to get your ACK/utACK :-)

File: src/script/script.cpp [hunk 7]
```diff
@@ -276,6 +282,11 @@ bool CScript::HasValidOps() const
     return true;
 }
 
+#ifdef _MSC_VER
+// Re-enable MSVC warnings after C4018 suppression in CScript::HasValidOps()
+#pragma warning(pop)
+#endif
```
Comment (MarcoFalke) [inline src/script/script.cpp]: NACK on adding these. It makes the code harder to parse as a human and they don't add any value.

**Comment by practicalswift** [comment]
@MarcoFalke Thanks for reviewing. Feedback addressed. Please re-review :-)

**Comment by NicolasDorier** [comment]
re-utACK for me

**Comment by sipsorcery** [comment]
tACK https://github.com/bitcoin/bitcoin/commit/bc7aff79fc0cfb10d25a7dd5928a1cb8c36a063b & https://github.com/bitcoin/bitcoin/pull/14151/commits/1423334fa0b45159c2304fda9e768d156e89f4a4.

**Comment by ken2812221** [comment]
ACK bc7aff79fc0cfb10d25a7dd5928a1cb8c36a063b

**Comment by Empact** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/14151/commits/bc7aff79fc0cfb10d25a7dd5928a1cb8c36a063b

File: src/test/crypto_tests.cpp [hunk 8]
```diff
@@ -532,7 +532,7 @@ BOOST_AUTO_TEST_CASE(countbits_tests)
             // Check handling of zero.
             BOOST_CHECK_EQUAL(CountBits(0), 0U);
         } else if (i < 10) {
-            for (uint64_t j = 1 << (i - 1); (j >> i) == 0; ++j) {
+            for (uint64_t j = static_cast<uint64_t>(1) << (i - 1); (j >> i) == 0; ++j) {
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/crypto_tests.cpp]: @laanwj `(uint64_t)1` is an alternative. Worth breaking four utACK:s for that?

[the remark this answers is not in the document]
Comment (laanwj) [reply src/test/crypto_tests.cpp]: @luke-jr suggested `uint64_t(1)`, that seems more C++ish without being verbose, and seems to work locally (I don't know if it avoids the warning)

[the remark this answers is not in the document]
Comment (sipa) [reply src/test/crypto_tests.cpp]: I generally prefer C-style casts for primitive types (due to syntactic overload).

[the remark this answers is not in the document]
Comment (laanwj) [reply src/test/crypto_tests.cpp]: > Really worth breaking four utACK:s for changing that? :-)

Maybe not, though I'd dread to see another PR which—for whatever reason—changes the syntax again later, so I'd rather have agreement on this on first go.

[the remark this answers is not in the document]
Comment (luke-jr) [reply src/test/crypto_tests.cpp]: `j` is at most 256, so this should work:

```C++
for (unsigned j = 1U << (i - 1); (j >> i) == 0; ++j) {
```

**Comment by practicalswift** [comment]
@laanwj Fixed! Sorry for the tag. Please re-review :-)

@NicolasDorier @sipsorcery @ken2812221 @Empact Would you mind re-reviewing?

**Comment by practicalswift** [comment]
@sipsorcery Is there an integer-suffix guaranteeing a width of exactly 64 bits on supported platforms? :-)

**Comment by sipsorcery** [comment]
> I don't think there exists an integer-suffix in C++11 guaranteeing a width of exactly 64 bits on all supported platforms? :-)

The C++17 specification states that `ul` can be represented by `unsigned long` or `unsigned long long` so no there is no guarantee of exactly 64 bits. It also indicates that `uint64_t`, `uint32_t` etc. are optional.

I've no argument either way.

**Comment by laanwj** [comment]
> Personally I would use an integer-suffix instead of the static cast, e.g. 0x0x08000000ul but it's purely a style thing and of no consequence here.

> @sipsorcery I don't think there exists an integer-suffix in C++11 guaranteeing a width of exactly 64 bits on all supported platforms? :-)

Yeah, indeed, if only C++ had Rust's `u64` suffix it would be easy!

**Comment by laanwj** [comment]
utACK b9dafe7d9ffcbe7928ffbfba816b54e196c57664
